include/mem_native.h

```diff
@@ -0,0 +1,115 @@
+/*
+ * SPDX-License-Identifier: GPL-2.0-or-later
+ *
+ * Copyright (C) 2020-2020  The dosbox-staging team
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#ifndef DOSBOX_MEM_NATIVE_H
+#define DOSBOX_MEM_NATIVE_H
+
+/*  These functions read and write 16, 32, and 64 bit uint's to
+ *  and from 8-bit memory positions using defined alignment. Use
+ *  these functions instead of C-style pointer type casts to larger
+ *  types, which are not alignment-safe.
+ */
+
+#include <cstdint>
+#include <cstring>
+
+/*  Read a uint16 from 8-bit native byte-ordered memory. Use this
+ *  instead of *(uint16_t*)(8_bit_ptr) or *(uint16_t*)(8_bit_ptr + offset)
+ */
+static inline uint16_t read_uint16(const uint8_t *arr)
+{
+	uint16_t val;
+	memcpy(&val, arr, sizeof(val));
+	return val;
+}
+
+/*  Read an array-indexed uint16 from 8-bit native byte-ordered memory.
+ *  Use this instead of ((uint16_t*)arr)[index]
+ */
+static inline uint16_t read_uint16_at(const uint8_t *arr, const uintptr_t index)
+{
+	return read_uint16(arr + index * sizeof(uint16_t));
+}
+
+// Write a uint16 to 8-bit memory using native byte-ordering.
+static inline void write_uint16(uint8_t *arr, uint16_t val)
+{
+	memcpy(arr, &val, sizeof(val));
+}
+
+// Write an array-indexed uint16 to 8-bit memory using native byte-ordering.
+static inline void write_uint16_at(uint8_t *arr, const uintptr_t index, const uint16_t val)
+{
+	write_uint16(arr + index * sizeof(uint16_t), val);
```

Comment:
Interesting, that GitHub highlights `index` word in here. That's because there is `index` function in `<strings.h>`. I suggest renaming this parameter to `idx` - it will be more in line with `arr` and `val`.

Reply:
Switched to idx across both files.